Adds two new switches to the maven test goal

hgschmie opened the pull request:
- skipTestExecution (boolean), default is false

If true, create the native image for the tests but do not execute it. This allows the creation of the native images for testing even if they don't fully execute / tests fail. Otherwise, the build would stop at the first test failure.

- failNoTests (boolean), default is true

If true, fail building if no tests were found. In multi-module builds, there are often modules that have no tests (documentation) or tests are enabled/disabled by profiles (slow tests, e2e tests etc.). This switch allows for a concise configuraiton of the native plugin and then skipping modules where no tests were run without failing the build.

hgschmie commented:
pretty ping?

hgschmie commented:
Added documentation, rebased to the current main branch.

dnestoro commented:
@hgschmie thanks for your patience. Can you please also provide test that shows how these flags work? 

You can add a new profile on the existing samples, where you will enable your flags, and add new test in the corresponding functional test. Check functional tests here: https://github.com/graalvm/native-build-tools/tree/master/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven

hgschmie commented:
Ok, will do tonight.

hgschmie commented:
@dnestoro ready to be reviewed.

dnestoro commented on native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/MavenTestExecutionFunctionalTests.groovy:
```diff
@@ -0,0 +1,71 @@
+        mvn '-DquickBuild', '-DfailNoTests=false', '-Pnative', 'test'
+
+        then:
+        buildSucceeded
```
Maybe add expected output that reflects your changes (from [here](https://github.com/graalvm/native-build-tools/pull/720/files#diff-407c562b048d82141f278e817ad7f1603db1be9c062a6e08816925d4358c0a89R175-R181))

hgschmie replied:
good callout! Done.

dnestoro commented on samples/non-native/pom.xml:
```diff
@@ -0,0 +1,167 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can't you use existing sample `java-application` which already doesn't have tests? See: https://github.com/graalvm/native-build-tools/tree/master/samples/java-application.

dnestoro replied:
But is that real scenario then? Why would someone have classes in `src/test` that doesn't have tests? I am okay with the change, but I am just wondering if it has real usecase?

hgschmie commented on docs/src/docs/asciidoc/end-to-end-maven-guide.adoc:
```diff
@@ -232,6 +232,76 @@ This works similarly to `JAVA_TOOL_OPTIONS`, where the value of the environment
 
 Learn more about Native Image build configuration https://www.graalvm.org/reference-manual/native-image/overview/BuildConfiguration/[on the website].
 
+==== Build, but not execute native tests
```
It would be useful to provide such guidance before asking for documentation. I am ok changing it, but I would like to go what the expected outcome is, otherwise it feels a bit like I am aiming at a moving target. I tried to follow the style in the maven-plugin.adoc file where there is a short, reference style explanation for each switch.

I shrank the text in the end-to-end guide down to a short paragraph and moved the bulk to the maven.adoc file in a new section "Controlling test execution".

hgschmie commented on native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/MavenTestExecutionFunctionalTests.groovy:
```diff
@@ -0,0 +1,71 @@
+
+class MavenTestExecutionFunctionalTests extends AbstractGraalVMMavenFunctionalTest {
+    def "test if the skipTestExecution flag skips execution of tests"() {
+        withSample("junit-tests")
```
done.

hgschmie commented on docs/src/docs/asciidoc/end-to-end-maven-guide.adoc:
```diff
@@ -232,6 +232,76 @@ This works similarly to `JAVA_TOOL_OPTIONS`, where the value of the environment
+
+This can be done by configuring the `<failNoTests>` configuration option for the full build and then overriding it in modules where the default behavior is not desired.
+
+If the majority of the submodules should contain native tests, nothing needs to be configured globally. This is the default for the native plugin. In sub-modules that do not contain native tests but should not fail the build, add this to the pom in the module in the `<plugins>` section:
```
btw, intellij whines here: 
<img width="428" height="74" alt="Screenshot 2025-09-25 at 18 27 06" src="https://github.com/user-attachments/assets/23b2c891-05d9-48b2-8265-f0a99b9c16fe" />

hgschmie replied:
I still changed it, though.

hgschmie commented on docs/src/docs/asciidoc/end-to-end-maven-guide.adoc:
```diff
@@ -232,6 +232,76 @@ This works similarly to `JAVA_TOOL_OPTIONS`, where the value of the environment
+
+==== Skip modules without native tests in multi-module builds
+
+Multi-module projects where only some sub-modules contain native tests need to skip test execution for any module that do not contain native tests. Any module that should contain tests, but does not, should fail the build.
```
done. Thanks, this sounds much more rounded.

hgschmie commented on docs/src/docs/asciidoc/end-to-end-maven-guide.adoc:
```diff
@@ -232,6 +232,76 @@ This works similarly to `JAVA_TOOL_OPTIONS`, where the value of the environment
 
+==== Build, but not execute native tests
+
+In a transition phase, it is often useful to build tests but not execute them. This allows inspection of the generated test image without failing the build because the tests do not execute successfully.
```
done

hgschmie commented on docs/src/docs/asciidoc/end-to-end-maven-guide.adoc:
```diff
@@ -232,6 +232,76 @@ This works similarly to `JAVA_TOOL_OPTIONS`, where the value of the environment
+</configuration>
+----
+
+and then override this in any sub-module that should have native tests by adding this to the module in the `<plugins>` section:
```
done.

hgschmie commented on docs/src/docs/asciidoc/maven-plugin.adoc:
```diff
@@ -61,6 +61,18 @@ The following configuration options are available:
 <skipNativeTests>true</skipNativeTests>
 ----
+`<skipTestExecution>`::
+   Skips execution of native tests. If set to true, the plugin will create the native image for the tests but do not execute the tests. This allows the creation of the native image for testing even if they don't fully execute / tests fail. Otherwise, the build stops at the first test failure. To skip only the execution of the native image compiled tests, add:
```
done

hgschmie commented:
addressed the doc comments, moved docs around to reduce the size of the end-to-end guide text.
updated the functional test; can not use java-application because it has not code at all in the src/test folder; I did not want to add some dummy class here because it may interfere with other functional tests. LMK how to proceed here. 

@dnestoro

hgschmie commented:
anything more needed from me? I feel that I addressed all issues and this is good to be merged. This has been open for a very long time.
